### PR TITLE
Fix Array#flatten when the element is T.untyped

### LINKED ISSUE #4033: Segmentation fault in `flatten` calls with empty sub array
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aa%20%3D%20%5B1%2C%202%2C%20%5B%5D%5D%0Aa.flatten)

```ruby
a = [1, 2, []]
a.flatten
```

#### Observed output

```
Segmentation fault (core dumped)
```

#### Expected behavior

No crash

- - -

Bisect pointed to this commit https://github.com/sorbet/sorbet/pull/4025/. We end up recursing forever and causing a stackoverflow.

### PR DESCRIPTION
**Pull request by elliottt**
### Motivation

Terminating the recursion of the `Array#flatten` intrinsic when the elements of the array are `T.untyped`.

Fixes #4033

### Test plan

See included automated tests.

### CODE REVIEW

File: core/types/calls.cc [hunk 1]
```diff
@@ -2572,6 +2572,10 @@ class Array_flatten : public IntrinsicMethod {
     // If the element type supports the #to_ary method, then Ruby will implicitly call it when flattening. So here we
     // dispatch #to_ary and recurse further down the result if it succeeds, otherwise we just return the type.
     static TypePtr typeToAry(const GlobalState &gs, const DispatchArgs &args, const TypePtr &type, const int newDepth) {
+        if (type.isUntyped()) {
```
Comment (Morriar) [inline core/types/calls.cc:2575]: Nice find 👍

**Review by jez — APPROVED** [review]

File: test/testdata/infer/array_flatten_T.untyped.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# typed: true
```
Comment (jez) [inline test/testdata/infer/array_flatten_T.untyped.rb:1]: `*.untyped.rb` 🙈

Comment (elliottt) [reply test/testdata/infer/array_flatten_T.untyped.rb:1]: `:shame:`

**Comment by kddeisz** [comment]
Thanks @elliottt for fixing this!

**Pull request merged**